userapi/src/main/java/fk/prof/userapi/cache/ZKBasedCacheInfoRegistry.java

```diff
@@ -0,0 +1,264 @@
+    private static final String distributedLockPath = "/global_mutex";
+
+    private final String zkNodesLoadInfoPath;
+    private final CuratorFramework curatorClient;
+    private final InterProcessLock interProcessLock;
+
+    private final AtomicReference<ConnectionState> connectionState;
+    private final AtomicReference<LocalDateTime> lastZkLostTime;
+    private final AtomicBoolean recentlyZkConnectionLost;
+
+    private final byte[] myResidencyInfoInBytes;
+    private final ProfileResidencyInfo myProfileResidencyInfo;
+    private Runnable cacheInvalidator;
+
+    private Counter lockAcquireTimeoutCounter = Util.counter("zk.lock.timeouts");
+    private Counter onReconnectFailures = Util.counter("zk.onreconnect.failures");
+
+    ZKBasedCacheInfoRegistry(CuratorFramework curatorClient, String myIp, int port, Runnable cacheInvalidator) {
+        this.curatorClient = curatorClient;
+
+        this.zkNodesLoadInfoPath = "/nodesInfo/" + myIp + ":" + port;
+        this.interProcessLock = new InterProcessMutex(curatorClient, distributedLockPath);
+        this.lastZkLostTime = new AtomicReference<>(LocalDateTime.MIN);
+        this.recentlyZkConnectionLost = new AtomicBoolean(false);
+        this.myProfileResidencyInfo = ProfileResidencyInfo.newBuilder().setIp(myIp).setPort(port).build();
+        this.myResidencyInfoInBytes = this.myProfileResidencyInfo.toByteArray();
+        this.curatorClient.getConnectionStateListenable().addListener(this::zkStateChangeListener,
+            Executors.newSingleThreadExecutor(new ThreadFactoryBuilder().setNameFormat("curator-state-listener").build()));
+
+        this.cacheInvalidator = cacheInvalidator;
+
+        this.connectionState = new AtomicReference<>(
+            curatorClient.getZookeeperClient().isConnected() ? ConnectionState.Connected : ConnectionState.Disconnected);
+    }
+
+    @Override
+    public void init() throws Exception {
+        ensureConnected();
+        ensureRequiredZkNodesPresent();
+    }
+
+    @Override
+    public void releaseOwnership(AggregatedProfileNamingStrategy profileName) throws Exception {
+        try(AutoCloseable ignored = getLock()) {
+            ProfileResidencyInfo residencyInfo = getProfileResidencyInfo(profileName);
+            if(residencyInfo != null && Objects.equals(residencyInfo, myProfileResidencyInfo)) {
+                removeProfileResidencyInfo(profileName);
+            }
+        }
+    }
+
+    @Override
+    public void claimOwnership(AggregatedProfileNamingStrategy profileName) throws Exception {
+        try (AutoCloseable ignored = getLock()){
+            ProfileResidencyInfo residencyInfo = getProfileResidencyInfo(profileName);
+            if (residencyInfo == null) {
+                putProfileResidencyInfo(profileName, false);
+            } else if(Objects.equals(residencyInfo, myProfileResidencyInfo) && sessionIdMatches(profileName)) {
```

Comment:
Did not understand. What does this achieve? If the current session is the one that created this node, then why do we delete and put again in zookeeper?